homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,251 @@
+    command_topic = config.get(mqtt.CONF_COMMAND_TOPIC)
+    state_topic = config.get(mqtt.CONF_STATE_TOPIC)
+    supported_feature_strings = config.get(ATTR_SUPPORTED_FEATURES)
+    supported_features = strings_to_services(supported_feature_strings)
+    qos = config.get(mqtt.CONF_QOS)
+    state_template = cv.template('{{value_json.battery_level}}')
+    state_template.hass = hass
+
+    add_devices([
+        MqttVacuum(name, supported_features, command_topic, state_topic, qos,
+                   state_template),
+    ])
+
+
+class MqttVacuum(VacuumDevice):
+    """Representation of a MQTT-controlled vacuum."""
+
+    # pylint: disable=no-self-use
+    def __init__(self, name, supported_features, command_topic, state_topic,
+                 qos, state_template):
+        """Initialize the vacuum."""
+        self._name = name
+        self._command_topic = command_topic
+        self._state_topic = state_topic
+        self._supported_features = supported_features
+        self._qos = qos
+
+        self._payload_turn_on = "turn_on"
+        self._payload_turn_off = "turn_off"
+        self._payload_return_to_base = "return_to_base"
+        self._payload_stop = "stop"
+        self._payload_clean_spot = "clean_spot"
+        self._payload_locate = "locate"
+        self._payload_start_pause = "start_pause"
+        self._retain = False
+
+        self._state_template = state_template
+
+        self._state = False
+        self._status = 'Unknown'
+        self._battery_level = 0
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe MQTT events.
+
+        This method is a coroutine.
+        """
+        @callback
+        def message_received(topic, payload, qos):
+            """Handle new MQTT message."""
+            try:
+                payload = json.loads(payload)
+            except ValueError:  # e.g. json.decoder.JSONDecodeError:
+                _LOGGER.warning("JSONDecodeError in vacuum.mqtt payload: %s",
+                                payload)
+                return
+
+            battery_level = payload.get(ATTR_BATTERY_LEVEL)
+            if battery_level is not None:
+                self._battery_level = int(battery_level)
+
+            charging = payload.get("charging")
+            state = payload.get(ATTR_STATE)
+            if payload[ATTR_STATE]:
+                if state == "cleaning":
+                    self._state = True
+                    self._status = "Cleaning"
+                if state == "docked":
+                    self._state = False
+                    if charging:
+                        self._status = "Docked & Charging"
+                    else:
+                        self._status = "Docked"
+                if state == "stopped":
+                    self._state = False
+                    self._status = "Stopped"
+
+            self.hass.async_add_job(self.async_update_ha_state())
+
+        yield from mqtt.async_subscribe(
```

Comment:
Please use the new style `self.hass.components.mqtt.async_....`

Reply:
will do!